x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/contextualai/ContextualAiService.java

```diff
@@ -0,0 +1,266 @@
+
+        return createModel(
+            inferenceEntityId,
+            taskType,
+            serviceSettingsMap,
+            taskSettingsMap,
+            secretSettingsMap,
+            ServiceUtils.parsePersistedConfigErrorMsg(inferenceEntityId, NAME),
+            ConfigurationParseContext.PERSISTENT
+        );
+    }
+
+    @Override
+    public ContextualAiRerankModel parsePersistedConfig(String inferenceEntityId, TaskType taskType, Map<String, Object> config) {
+        Map<String, Object> serviceSettingsMap = ServiceUtils.removeFromMapOrThrowIfNull(config, ModelConfigurations.SERVICE_SETTINGS);
+        Map<String, Object> taskSettingsMap = ServiceUtils.removeFromMapOrDefaultEmpty(config, ModelConfigurations.TASK_SETTINGS);
+
+        return createModel(
+            inferenceEntityId,
+            taskType,
+            serviceSettingsMap,
+            taskSettingsMap,
+            null,
+            ServiceUtils.parsePersistedConfigErrorMsg(inferenceEntityId, NAME),
+            ConfigurationParseContext.PERSISTENT
+        );
+    }
+
+    @Override
+    public void doInfer(
+        Model model,
+        InferenceInputs inputs,
+        Map<String, Object> taskSettings,
+        TimeValue timeout,
+        ActionListener<InferenceServiceResults> listener
+    ) {
+        if (model instanceof ContextualAiRerankModel == false) {
+            listener.onFailure(ServiceUtils.createInvalidModelException(model));
+            return;
+        }
+
+        ContextualAiRerankModel contextualAiModel = (ContextualAiRerankModel) model;
+        var actionCreator = new ContextualAiActionCreator(getSender(), getServiceComponents());
+
+        var action = contextualAiModel.accept(actionCreator, taskSettings);
+        action.execute(inputs, timeout, listener);
+    }
+
+    @Override
+    protected void doChunkedInfer(
+        Model model,
+        EmbeddingsInput inputs,
+        Map<String, Object> taskSettings,
+        InputType inputType,
+        TimeValue timeout,
+        ActionListener<List<ChunkedInference>> listener
+    ) {
+        listener.onFailure(new ElasticsearchStatusException("Chunked inference is not supported for rerank task", RestStatus.BAD_REQUEST));
+    }
+
+    @Override
+    protected void doUnifiedCompletionInfer(
+        Model model,
+        UnifiedChatInput inputs,
+        TimeValue timeout,
+        ActionListener<InferenceServiceResults> listener
+    ) {
+        listener.onFailure(new ElasticsearchStatusException("Unified completion is not supported for rerank task", RestStatus.BAD_REQUEST));
+    }
+
+    @Override
+    protected void validateInputType(InputType inputType, Model model, ValidationException validationException) {
+        // Rerank accepts any input type
+    }
+
+    @Override
+    public int rerankerWindowSize(String modelId) {
+        // Using conservative default as the actual window size is not known
+        return RerankingInferenceService.CONSERVATIVE_DEFAULT_WINDOW_SIZE;
```

Comment:
The purpose of this method is to define the max size of text chunks that can be sent to the reranker. Long documents are automatically chunked for the reranker. `CONSERVATIVE_DEFAULT_WINDOW_SIZE` is a small value that will fit into the smallest context windows.


The Contextual AI rerank [docs](https://docs.contextual.ai/api-reference/rerank/rerank) state "The total request cannot exceed 400,000 tokens. The combined length of the query, instruction and any document with its metadata must not exceed 8,000 tokens."

The max is 8,000 tokens. The function wants the number of words and that depends on the tokenizer used. I'm guessing this value should be in the range of 5000 to 6000 words.